Mathlib/GroupTheory/Submonoid/Operations.lean

```diff
@@ -0,0 +1,1523 @@
+#align submonoid.map_sup_comap_of_surjective Submonoid.map_sup_comap_of_surjective
+#align add_submonoid.map_sup_comap_of_surjective AddSubmonoid.map_sup_comap_of_surjective
+
+@[to_additive]
+theorem map_supᵢ_comap_of_surjective (S : ι → Submonoid N) : (⨆ i, (S i).comap f).map f = supᵢ S :=
+  (giMapComap hf).l_supᵢ_u _
+#align submonoid.map_supr_comap_of_surjective Submonoid.map_supᵢ_comap_of_surjective
+#align add_submonoid.map_supr_comap_of_surjective AddSubmonoid.map_supᵢ_comap_of_surjective
+
+@[to_additive]
+theorem comap_le_comap_iff_of_surjective {S T : Submonoid N} : S.comap f ≤ T.comap f ↔ S ≤ T :=
+  (giMapComap hf).u_le_u_iff
+#align submonoid.comap_le_comap_iff_of_surjective Submonoid.comap_le_comap_iff_of_surjective
+#align add_submonoid.comap_le_comap_iff_of_surjective AddSubmonoid.comap_le_comap_iff_of_surjective
+
+@[to_additive]
+theorem comap_strict_mono_of_surjective : StrictMono (comap f) :=
+  (giMapComap hf).strictMono_u
+#align submonoid.comap_strict_mono_of_surjective Submonoid.comap_strict_mono_of_surjective
+#align add_submonoid.comap_strict_mono_of_surjective AddSubmonoid.comap_strict_mono_of_surjective
+
+end GaloisInsertion
+
+end Submonoid
+
+namespace OneMemClass
+
+variable {A M₁ : Type _} [SetLike A M₁] [One M₁] [hA : OneMemClass A M₁] (S' : A)
+
+/-- A submonoid of a monoid inherits a 1. -/
+@[to_additive "An `AddSubmonoid` of an `AddMonoid` inherits a zero."]
+instance one : One S' :=
+  ⟨⟨1, OneMemClass.one_mem S'⟩⟩
+#align one_mem_class.has_one OneMemClass.one
+#align zero_mem_class.has_zero ZeroMemClass.zero
+
+@[simp, norm_cast, to_additive]
+theorem coe_one : ((1 : S') : M₁) = 1 :=
+  rfl
+#align one_mem_class.coe_one OneMemClass.coe_one
+#align zero_mem_class.coe_zero ZeroMemClass.coe_zero
+
+variable {S'}
+
+@[simp, norm_cast, to_additive]
+theorem coe_eq_one {x : S'} : (↑x : M₁) = 1 ↔ x = 1 :=
+  (Subtype.ext_iff.symm : (x : M₁) = (1 : S') ↔ x = 1)
+#align one_mem_class.coe_eq_one OneMemClass.coe_eq_one
+#align zero_mem_class.coe_eq_zero ZeroMemClass.coe_eq_zero
+
+variable (S')
+
+@[to_additive]
+theorem one_def : (1 : S') = ⟨1, OneMemClass.one_mem S'⟩ :=
+  rfl
+#align one_mem_class.one_def OneMemClass.one_def
+#align zero_mem_class.zero_def ZeroMemClass.zero_def
+
+end OneMemClass
+
+variable {A : Type _} [SetLike A M] [hA : SubmonoidClass A M] (S' : A)
+
+/-- An `AddSubmonoid` of an `AddMonoid` inherits a scalar multiplication. -/
+instance AddSubmonoidClass.nSMul {M} [AddMonoid M] {A : Type _} [SetLike A M]
+    [AddSubmonoidClass A M] (S : A) : SMul ℕ S :=
+  ⟨fun n a => ⟨n • a.1, smul_mem a.2 n⟩⟩
+#align add_submonoid_class.has_nsmul AddSubmonoidClass.nSMul
+
+namespace SubmonoidClass
+
+/-- A submonoid of a monoid inherits a power operator. -/
+instance nPow {M} [Monoid M] {A : Type _} [SetLike A M] [SubmonoidClass A M] (S : A) : Pow S ℕ :=
+  ⟨fun a n => ⟨a.1 ^ n, pow_mem a.2 n⟩⟩
+#align submonoid_class.has_pow SubmonoidClass.nPow
+
+attribute [to_additive] nPow
+
```

Comment:
I guess there's a missing `#align` statement here...?

Reply:
I think my problem is that I don't fully know what is even happening here. We manually define both ```AddSubmonoidClass.nSMul``` and ```SubmonoidClass.nPow```, but we still need to tell lean that they correspond to one another via this ```to_additive``` attribute? And this should be functionally the same as adding ```@[to_additive]``` to the declaration of ```SubmonoidClass.nPow````, right? 

If that's correct, then I don't think we need an #align here, since we are already aligning the only possible thing we could be aligning here, namely
```lean
#align add_submonoid_class.has_nsmul AddSubmonoidClass.nSMul.
``` 

I won't be here for the rest of the day, so if there is a meaningful align statement that I don't see here, feel free to add it on my behalf!

Reply:
Yes, I agree with Lukas's analysis.